src/InjectHostDevBufferCopies.cpp

```diff
@@ -722,6 +732,7 @@ class InjectBufferCopiesForInputsAndOutputs : public IRMutator {
         void include(const Parameter &p) {
             if (p.defined()) {
                 result.insert(p.name());
+                result_storage[p.name()] = p.memory_type();
```

Comment:
OK, so the cause of the crash for tutorial_lesson_12 is that the `include(Buffer<>)` method also need to fill in entries for `result_storage`, eg

```
        void include(const Buffer<> &b) {
            if (b.defined()) {
                result.insert(b.name());
                result_storage[b.name()] = /* VALUE */;
            }
        }
```

...because as it is now, the code in the `mutate()` method below can try to find an entry that has no value here.

I don't know what the right value is, though; `Buffer<>` doesn't have a MemoryType andI don't know if `Heap` is right. Hmm...

Reply:
Addressed, I chose MemoryType::Auto, because:

* That's the default I have for Param
* It seems like Buffer<> mainly appears in the AST if you are declaring a Func in the inline/JIT syntax (correct?).  In that case you can always wrap it with ImageParam if you want to be able to override the memory type.